apps/sync/js/sync-engine/syncengine.js

```diff
@@ -0,0 +1,274 @@
+  var FxSyncIdSchema = Kinto.createIdSchema({
+    constructor: function(collectionName) {
+      this.collectionName = collectionName;
+    },
+    generate: function() {
+      var bytes = new Uint8Array(9);
+      crypto.getRandomValues(bytes);
+      var binStr = '';
+      for (var i=0; i<9; i++) {
+          binStr += String.fromCharCode(bytes[i]);
+      }
+      // See https://docs.services.mozilla.com/storage/apis-1.5.html
+      return window.btoa(binStr).replace('+', '-').replace('/', '_');
+    },
+    validate: function(id) {
+      // FxSync id's should be 12 ASCII characters, representing 9 bytes of data
+      // in modified Base64 for URL variants exist, where the '+' and '/'
+      // characters of standard Base64 are respectively replaced by '-' and '_'
+      // See https://docs.services.mozilla.com/storage/apis-1.5.html
+      return /^[A-Za-z0-9-_]{12}$/.test(id);
+    }
+  });
+
+  var ControlCollectionIdSchema = Kinto.createIdSchema({
+    constructor: function(collectionName, keyName) {
+      this.collectionName = collectionName;
+      this.keyName = keyName;
+    },
+    generate: function() {
+      return this.keyName;
+    },
+    validate: function(id) {
+      return (id === this.keyName);
+    }
+  });
+
+  var WebCryptoTransformer = Kinto.createRemoteTransformer({
+    constructor: function(collectionName, fswc) {
+      if (!fswc.bulkKeyBundle) {
+        throw new Error(`Attempt to register Transformer with no bulk key bundl\
+e!`);
+      }
+      this.collectionName = collectionName;
+      this.fswc = fswc;
+    },
+    encode: function(record) {
+      return this.fswc.encrypt(record.payload, this.collectionName)
+      .then(payloadEnc => {
+        record.payload = JSON.stringify(payloadEnc);
+        return record;
+      });
+    },
+    decode: function(record) {
+      // Allowing JSON.parse errors to bubble up to the errors list in the
+      // syncResults:
+      return this.fswc.decrypt(JSON.parse(record.payload), this.collectionName)
+      .then(payloadDec => {
+        record.payload = payloadDec;
+        return record;
+      });
+    }
+  });
+
+  var SyncEngine = function(options) {
+    if (typeof options !== 'object') {
+      throw new Error('options should be an Object');
+    }
+    if (typeof options.URL !== 'string') {
+      throw new Error('options.URL should be a String');
+    }
+    if (typeof options.assertion !== 'string') {
+      throw new Error('options.assertion should be a String');
+    }
+    if (typeof options.xClientState !== 'string') {
+      throw new Error('options.xClientState should be a String');
+    }
+    if (typeof options.kB !== 'string') {
+      throw new Error('options.kB should be a String');
+    }
+
+    this._kB = options.kB;
+    this._collections = {};
+    this._controlCollections = {};
+    this._adapters = {};
+    this._fswc = new FxSyncWebCrypto();
+    this._initKinto({
+       URL: options.URL,
+       assertion: options.assertion,
```

Comment:
Yes, kinto.js does not provide a way to update the options of an existing Kinto object, so it will be necessary to construct a new Kinto object. The simplest way to do this is to construct a whole new SyncEngine object. So in the [dbPrefix follow-up bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1202627#c2), I'll optimize the SyncEngine constructor so that it doesn't fetch meta/global and crypto/keys if a local copy of them are already available for the current user.